src/main/kotlin/org/rust/lang/refactoring/generateConstructor/GenerateConstructorHandler.kt

```diff
@@ -0,0 +1,44 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.refactoring.generateConstructor
+
+
+import com.intellij.codeInsight.CodeInsightActionHandler
+import com.intellij.codeInsight.actions.CodeInsightAction
+import com.intellij.lang.LanguageCodeInsightActionHandler
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RsFile
+import org.rust.lang.core.psi.RsStructItem
+import org.rust.lang.core.psi.ext.ancestorOrSelf
+
+class GenerateConstructorAction : CodeInsightAction() {
+    override fun getHandler(): CodeInsightActionHandler = GenerateConstructorHandler()
+    override fun isValidForFile(project: Project, editor: Editor, file: PsiFile): Boolean {
+        return GenerateConstructorHandler().isValidFor(editor, file)
+    }
+}
+
+class GenerateConstructorHandler : LanguageCodeInsightActionHandler {
+
+    override fun isValidFor(editor: Editor, file: PsiFile): Boolean {
```

Comment:
`isValidFor` and `invoke` contains the same code which should be extracted like this:
```
class GenerateConstructorHandler : LanguageCodeInsightActionHandler {
    override fun isValidFor(editor: Editor, file: PsiFile): Boolean =
        getStructItem(editor, file) != null

    override fun startInWriteAction(): Boolean = false

    override fun invoke(project: Project, editor: Editor, file: PsiFile) {
        val structItem = getStructItem(editor, file) ?: return
        generateConstructorBody(structItem, editor)
    }

    private fun getStructItem(editor: Editor, file: PsiFile): RsStructItem? =
        file.findElementAt(editor.caretModel.offset)?.ancestorOrSelf()
}
```